plugins/org.jkiss.dbeaver.net.ssh/src/org/jkiss/dbeaver/model/net/ssh/SSHUtils.java

```diff
@@ -78,4 +90,94 @@ public static boolean isKeyEncrypted(byte[] privKeyValue) {
+            } catch (InvalidPathException e) {
+                log.warn("Failed to resolve SSH known hosts file location at " + sshHomePathString, e);
+                if (forceFileObjectOnFail) {
+                    return new File(sshHomePathString + File.pathSeparator + KNOWN_SSH_HOSTS_FILE_NAME);
+                } else {
+                    return null;
+                }
+            }
+        } else {
+            // seems preference path not set at all, so try to resolve it and preserve
+            return resolveDefaultKnownSshHostsFile(forceFileObjectOnFail, true);
+        }
+    }
+
+    private static File resolveDefaultKnownSshHostsFile(boolean forceFileObjectOnFail, boolean updatePreferences) {
+        try {
+            String userHomePathString = System.getProperty(IConstants.SYSTEM_PROPERTY_USER_HOME);
+            if (userHomePathString != null) {
+                Path userHomeDirPath = Paths.get(userHomePathString);
+                if (userHomeDirPath.toFile().isDirectory()) {
+                    Path sshHomeDirPath = userHomeDirPath.resolve(DEFAULT_SSH_HOME_DIR_NAME);
+                    File sshHomeDir = sshHomeDirPath.toFile();
+
+                    if (RuntimeUtils.isWindows() && (!sshHomeDir.isDirectory() || !sshHomeDir.exists())) {
+                        Path sshHomeOldDirPath = userHomeDirPath.resolve(DEFAULT_SSH_HOME_DIR_NAME_WIN_OLD);
+                        File sshHomeOldDir = sshHomeDirPath.toFile();
+                        if (sshHomeOldDir.isDirectory()) {
+                            sshHomeDirPath = sshHomeOldDirPath;
+                            sshHomeDir = sshHomeOldDir;
+                        }
+                    }
+
+                    if (sshHomeDir.isDirectory() || !sshHomeDir.exists()) {
+                        // don't need to create it until we'll need to write known hosts file
+
+                        if (updatePreferences) {
+                            Platform.getPreferencesService()
+                                    .getRootNode().node(PLATFORM_SSH_PREFERENCES_NODE)
+                                    .put(PLATFORM_SSH_PREFERENCES_SSH2HOME_KEY, sshHomeDir.getAbsolutePath());
+                        }
+
+                        Path knownSshHostsFilePath = sshHomeDirPath.resolve(KNOWN_SSH_HOSTS_FILE_NAME);
+                        return knownSshHostsFilePath.toFile();
+                    } else {
+                        log.warn("Failed to resolve default SSH known hosts file location due to invalid SSH home directory " + sshHomeDirPath.toAbsolutePath());
+                    }
+                } else {
+                    log.warn("Failed to resolve default SSH known hosts file location due to missing user home directory " + userHomeDirPath.toAbsolutePath());
+                }
+            } else {
+                log.warn("Failed to resolve default SSH known hosts file location due to missing user home system property.");
+            }
```

Comment:
I think these duplicated logging lines can be moved just above line 170. If we get so far, there will be an error anyways.

Reply:
I suppose it's important to have meaningful error message which is different for these situations.
Anyway the point that we can don't get an error here, but something goes wrong. The error can be further in usage, but It can be avoided if user reads logs. So, I'm not sure that it's good to replace different messages with one not straight pointing to the problem.

Reply:
I haven't noticed that they aren't the same. My bad.